### PR TITLE
coverage: shard coverage tests

### PR DESCRIPTION
**Pull request by lizan**
Signed-off-by: Lizan Zhou <lizan@tetrate.io>

Description:
Risk Level: Low
Testing:
Docs Changes:
Release Notes:

### CODE REVIEW

File: test/coverage/gen_build.sh [hunk 1]
```diff
@@ -67,10 +67,12 @@ EOF
...
     ],
-    tags = ["manual"],
+    # no-remote due to https://github.com/bazelbuild/bazel/issues/4685
+    tags = ["manual", "no-remote"],
     coverage = False,
-    # Needed when invoking external shell tests etc.
-    local = True,
+    # Using the maximum sharding to allow Bazel to schedule them across CPUs.
+    # Sharding can be disabled by --test_sharding_strategy=disabled.
+    shard_count = 50,
```
Comment (htuch) [inline test/coverage/gen_build.sh]: Should this be a function of the number of local CPUs?

File: test/run_envoy_bazel_coverage.sh [hunk 2]
```diff
@@ -36,7 +36,10 @@ mkdir -p "${COVERAGE_DIR}"
 
 COVERAGE_IGNORE_REGEX="(/external/|pb\.(validate\.)?(h|cc)|/chromium_url/|/test/|/tmp)"
 COVERAGE_BINARY="bazel-bin/test/coverage/coverage_tests"
-COVERAGE_DATA="bazel-out/k8-fastbuild/testlogs/test/coverage/coverage_tests/coverage.dat"
+COVERAGE_DATA="${COVERAGE_DIR}/coverage.dat"
+
+echo "Merging coverage data..."
+llvm-profdata merge -sparse -o ${COVERAGE_DATA} $(find -L bazel-out/k8-fastbuild/testlogs/test/coverage/coverage_tests/ -name coverage.dat)
```
Comment (htuch) [inline test/run_envoy_bazel_coverage.sh]: Do you know how well this works when rebuilding locally after making a change and not cleaning the Bazel cache?

File: test/coverage/gen_build.sh (hunk 1 above)
Comment (lizan) [reply test/coverage/gen_build.sh]: No, I noted the reason in the comment. Unlike other places that we shard tests, the shards are very uneven, the single test from //test/integration:hotrestart_test takes > 100s while other tests are very small, sharding to maximum allow bazel to schedule the time taking shards run in one core while the other core may run tens shard that only takes ~10s.

File: test/run_envoy_bazel_coverage.sh (hunk 2 above)
Comment (lizan) [reply test/run_envoy_bazel_coverage.sh]: I'm not aware of any issue, coverage tests are no longer run out of sandbox and bazel is aware of the coverage.dat.

**Comment by lizan** [comment]
Yeah in Circle is 40mins vs 30mins, the test running time is down from 10mins to 1min.

File: test/coverage/gen_build.sh (hunk 1 above)
Comment (htuch) [reply test/coverage/gen_build.sh]: Can you say this more explicitly in the comment?

File: test/run_envoy_bazel_coverage.sh (hunk 2 above)
Comment (htuch) [reply test/run_envoy_bazel_coverage.sh]: Great, being back in the sandbox is a huge win when reasoning about this kind of thing.

**Comment by htuch** [comment]
Thanks!

**Pull request merged**